### PR TITLE
Multicallable

### PR DESCRIPTION
**Pull request by clabby**
# Overview

Adds an implementation of @Vectorized's [Multicallable](https://github.com/vectorized/solady/blob/main/src/utils/Multicallable.sol).

## Checklist
- [x] Finish impl
- [x] Port Solady test suite
- [ ] Optimize

### CODE REVIEW

**Comment by Vectorized** [comment]
I’ve merged it into Solady

File: src/utils/Multicallable.huff [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+#define constant DATA_OFFSET = 0x44
+// Memory
+#define constant RES = 0x20
+#define constant RES_OFF = 0x40
+
+/// @notice Macro that enables a single call to call multiple methods within its contract.
+/// @author Solady (https://github.com/vectorized/solady/blob/main/src/utils/Multicallable.sol)
+/// @author Modified from Solmate (https://github.com/transmissions11/solmate/blob/main/src/utils/Multicallable.sol)
+/// @author clabby <https://github.com/clabby>
+#define macro MULTICALL() = takes (0) returns (0) {
```
Comment (abigger87) [inline src/utils/Multicallable.huff:12]: Should we conform to the natspec at top of contract? (we should probably also add to huff-docs style guide fwiw)

File: src/utils/Multicallable.huff [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        dup2 0x20 add  // [o + 0x20, cur_bytes_len, o, data_end, mem_ptr, results_offset]
+        dup5           // [mem_ptr, o + 0x20, cur_bytes_len, o, data_end, mem_ptr, results_offset]
+        calldatacopy   // [o, data_end, mem_ptr, results_offset]
+
+        calldataload   // [cur_bytes_len, data_end, mem_ptr, results_offset]
+        0x00 dup1      // [0x00, 0x00, cur_bytes_len, data_end, mem_ptr, results_offset]
+        swap2          // [cur_bytes_len, 0x00, 0x00, data_end, mem_ptr, results_offset]
+        dup5           // [mem_ptr, cur_bytes_len, 0x00, 0x00, data_end, mem_ptr, results_offset]
+        address        // [self_addr, mem_ptr, cur_bytes_len, 0x00, 0x00, data_end, mem_ptr, results_offset]
+        gas            // [gas, self_addr, mem_ptr, cur_bytes_len, 0x00, 0x00, data_end, mem_ptr, results_offset]
```
Comment (abigger87) [inline src/utils/Multicallable.huff:64]: tg delegatecall returns leftover gas 😅

**Review by abigger87 — CHANGES_REQUESTED** [review]
Potential Nit, otherwise really nice work!

File: src/utils/Multicallable.huff (hunk 1 above)
Comment (clabby) [reply src/utils/Multicallable.huff]: Yeah- I think that's a good idea. Will throw up an issue in the docs, fixed it :smile:

**Review by abigger87 — APPROVED** [review]
LGTM

**Pull request merged**